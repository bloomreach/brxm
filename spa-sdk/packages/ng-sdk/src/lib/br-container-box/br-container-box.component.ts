/*
 * Copyright 2020 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { ChangeDetectionStrategy, Component, HostBinding, Input } from '@angular/core';
import { Container, Page } from '@bloomreach/spa-sdk';

@Component({
  selector: 'div.br-container-box', // tslint:disable-line: component-selector
  templateUrl: './br-container-box.component.html',
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class BrContainerBoxComponent {
  @Input() component!: Container;
  @Input() page!: Page;

  @HostBinding('class.hst-container')
  get isPreview() {
    return this.page?.isPreview();
  }
}
