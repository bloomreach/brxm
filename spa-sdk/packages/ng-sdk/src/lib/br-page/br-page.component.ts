/*
 * Copyright 2020 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  AfterContentChecked,
  AfterViewInit,
  ChangeDetectorRef,
  ChangeDetectionStrategy,
  Component,
  ContentChild,
  Inject,
  Input,
  OnChanges,
  OnDestroy,
  OnInit,
  Output,
  Optional,
  PLATFORM_ID,
  SimpleChanges,
  TemplateRef,
  Type,
  ViewChild,
} from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { isPlatformBrowser, isPlatformServer } from '@angular/common';
import { makeStateKey, StateKey, TransferState } from '@angular/platform-browser';
import { from, of, BehaviorSubject, ReplaySubject, Subject } from 'rxjs';
import { filter, map, mapTo, pairwise, pluck, switchMap, take } from 'rxjs/operators';
import { destroy, initialize, isPage, Configuration, Page, PageModel } from '@bloomreach/spa-sdk';
import { BrComponentContext } from '../br-component.directive';
import { BrProps } from '../br-props.model';

interface BrNodeContext extends BrComponentContext {
  template?: TemplateRef<BrComponentContext>;
}

/**
 * The brXM page.
 */
@Component({
  changeDetection: ChangeDetectionStrategy.OnPush,
  selector: 'br-page',
  templateUrl: './br-page.component.html',
})
export class BrPageComponent implements AfterContentChecked, AfterViewInit, OnChanges, OnDestroy {
  /**
   * The configuration of the SPA SDK.
   * @see https://www.npmjs.com/package/@bloomreach/spa-sdk#configuration
   */
  @Input() configuration!: Omit<Configuration, 'httpClient'>;

  /**
   * The brXM and Angular components mapping.
   */
  @Input() mapping: Record<string, Type<BrProps>> = {};

  /**
   * The pre-initialized page instance or prefetched page model.
   * Mostly this property should be used to transfer state from the server-side to the client-side.
   */
  @Input() page?: Page | PageModel;

  /**
   * The TransferState key is used to transfer the state from the server-side to the client-side.
   * By default, it equals to `brPage`.
   * If `false` is passed then the state transferring feature will be disabled.
   */
  @Input() stateKey: StateKey<PageModel | undefined> | false = makeStateKey('brPage');

  /**
   * The current state of the page component.
   */
  @Output() state = new BehaviorSubject<Page | undefined>(undefined);

  @ViewChild('brNode') node!: TemplateRef<BrNodeContext>;

  @ContentChild(TemplateRef) private template?: TemplateRef<BrComponentContext>;

  private afterContentChecked$ = new Subject();

  private afterViewInit$ = new ReplaySubject();

  constructor(
    private changeDetectorRef: ChangeDetectorRef,
    private httpClient: HttpClient,
    @Inject(PLATFORM_ID) private platform: any,
    @Optional() private transferState?: TransferState,
  ) {
    this.request = this.request.bind(this);

    this.state.pipe(
      pairwise(),
      pluck(0),
      filter(isPage),
    )
    .subscribe(destroy);

    this.state.pipe(
      filter(isPage),
      switchMap((page) => this.afterContentChecked$.pipe(take(1), mapTo(page))),
    )
    .subscribe((page) => page.sync());

    this.state.pipe(
      filter(() => isPlatformServer(this.platform)),
      filter(isPage),
    )
    .subscribe((page) => this.stateKey && this.transferState?.set(this.stateKey, page.toJSON()));
  }

  get context(): BrNodeContext | undefined {
    const page = this.state.getValue();
    const component = page?.getComponent();

    if (!page || !component) {
      return;
    }

    return {
      component,
      page,
      $implicit: component,
      template: this.template,
    };
  }

  ngOnChanges(changes: SimpleChanges): void {
    if (changes.configuration || changes.page) {
      this.initialize(changes.page?.currentValue);
    }

    if (changes.stateKey?.previousValue && isPlatformServer(this.platform)) {
      if (changes.stateKey.currentValue && this.transferState?.hasKey(changes.stateKey.previousValue)) {
        this.transferState?.set(
          changes.stateKey.currentValue,
          this.transferState?.get(changes.stateKey.previousValue, undefined),
        );
      }

      this.transferState?.remove(changes.stateKey.previousValue);
    }
  }

  ngAfterContentChecked(): void {
    this.afterContentChecked$.next();
  }

  ngAfterViewInit(): void {
    this.afterViewInit$.next();
  }

  ngOnDestroy(): void {
    this.state.next(undefined);
    this.state.complete();
    this.afterContentChecked$.complete();
    this.afterViewInit$.complete();
  }

  private initialize(page: Page | PageModel | undefined): void {
    if (this.stateKey && isPlatformBrowser(this.platform) && this.transferState?.hasKey(this.stateKey)) {
      page = page ?? this.transferState?.get(this.stateKey, undefined);
      this.transferState?.remove(this.stateKey);
    }

    const configuration = { httpClient: this.request, ...this.configuration } as Configuration;
    const observable = page
      ? of(initialize(configuration, page))
      : from(initialize(configuration));

    observable.pipe(
      switchMap(state => this.afterViewInit$.pipe(take(1), mapTo(state))),
    )
    .subscribe(state => {
      this.state.next(state);
      this.changeDetectorRef.detectChanges();
    });
  }

  private request(...[{ data: body, headers, method, url }]: Parameters<Configuration['httpClient']>) {
    return this.httpClient.request<PageModel>(
      method,
      url,
      {
        body,
        headers: headers as Record<string, string | string[]>,
        responseType: 'json'
      },
    )
    .pipe(map(data => ({data})))
    .toPromise();
  }
}
