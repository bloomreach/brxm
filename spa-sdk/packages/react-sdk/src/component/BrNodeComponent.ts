/*
 * Copyright 2019 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { Component } from '@bloomreach/spa-sdk';
import { BrProps } from './BrProps';
import { BrMappingContext } from './BrMappingContext';

export class BrNodeComponent<T extends Component> extends React.Component<BrProps<T>> {
  static contextType = BrMappingContext;
  context!: React.ContextType<typeof BrMappingContext>;

  protected getMapping(): string | undefined {
    return this.props.component.getName();
  }

  protected fallback() {
    return this.props.children;
  }

  render() {
    const mapping = this.getMapping();
    const component = mapping && this.context[mapping];
    if (!component) {
      return this.fallback();
    }

    return React.createElement(component, this.props);
  }
}
