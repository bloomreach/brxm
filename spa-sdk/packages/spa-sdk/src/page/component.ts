/*
 * Copyright 2019-2020 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { inject, injectable } from 'inversify';
import { LinkFactory } from './link-factory';
import { Link } from './link';
import { MetaCollectionFactory } from './meta-collection-factory';
import { MetaCollectionModel, MetaCollection } from './meta-collection';
import { Reference } from './reference';

export const ComponentChildrenToken = Symbol.for('ComponentChildrenToken');
export const ComponentModelToken = Symbol.for('ComponentModelToken');

/**
 * Generic component type.
 */
export const TYPE_COMPONENT = 'component';

/**
 * Container item type.
 */
export const TYPE_COMPONENT_CONTAINER_ITEM = 'container-item';

/**
 * Container type.
 */
export const TYPE_COMPONENT_CONTAINER = 'container';

export type ComponentType = typeof TYPE_COMPONENT
  | typeof TYPE_COMPONENT_CONTAINER_ITEM
  | typeof TYPE_COMPONENT_CONTAINER;

/**
 * Meta-data of a component.
 */
export interface ComponentMeta extends MetaCollectionModel {
  params?: ComponentParameters;
}

type ComponentLinks = 'self';

type ComponentModels = Record<string, any>;

type ComponentParameters = Partial<Record<string, string>>;

/**
 * Model of a component.
 */
export interface ComponentModel {
  children?: Reference[];
  id: string;
  links: Record<ComponentLinks, Link>;
  meta: ComponentMeta;
  models?: ComponentModels;
  name?: string;
  type: ComponentType;
}

/**
 * A component in the current page.
 */
export interface Component {
  /**
   * @return The component id.
   */
  getId(): string;

  /**
   * @return The component meta-data collection.
   */
  getMeta(): MetaCollection;

  /**
   * @return The map of the component models.
   */
  getModels<T extends ComponentModels>(): T;

  /**
   * @return The link to the partial component model.
   */
  getUrl(): string | undefined;

  /**
   * @return The name of the component.
   */
  getName(): string;

  /**
   * @return The parameters of the component.
   */
  getParameters(): ComponentParameters;

  /**
   * @return The direct children of the component.
   */
  getChildren(): Component[];

  /**
   * Looks up for a nested component.
   * @param componentNames A lookup path.
   */
  getComponent<U extends Component>(...componentNames: string[]): U | undefined;
  getComponent(): this;

  /**
   * Looks up for a nested component by its id.
   * @param id A component id.
   */
  getComponentById<U extends Component>(id: string): U | this | undefined;
}

@injectable()
export class ComponentImpl implements Component {
  protected meta: MetaCollection;

  constructor(
    @inject(ComponentModelToken) protected model: ComponentModel,
    @inject(ComponentChildrenToken) protected children: Component[],
    @inject(LinkFactory) private linkFactory: LinkFactory,
    @inject(MetaCollectionFactory) metaFactory: MetaCollectionFactory,
  ) {
    this.meta = metaFactory(this.model.meta);
  }

  getId() {
    return this.model.id;
  }

  getMeta() {
    return this.meta;
  }

  getModels<T extends ComponentModels>(): T;
  getModels() {
    return this.model.models || {};
  }

  getUrl() {
    return this.linkFactory.create(this.model.links.self);
  }

  getName() {
    return this.model.name || '';
  }

  getParameters() {
    return this.model.meta.params || {};
  }

  getChildren() {
    return this.children;
  }

  getComponent(): this;
  getComponent<U extends Component>(...componentNames: string[]): U | undefined;
  getComponent(...componentNames: string[]) {
    // tslint:disable-next-line:no-this-assignment
    let component: Component | undefined = this;

    while (componentNames.length && component) {
      const name = componentNames.shift()!;
      component = component.getChildren().find(component => component.getName() === name);
    }

    return component;
  }

  getComponentById<U extends Component>(id: string): U | this | undefined;
  getComponentById(id: string) {
    const queue = [this as Component];

    while (queue.length) {
      const component = queue.shift()!;

      if (component.getId() === id) {
        return component;
      }

      queue.push(...component.getChildren());
    }
  }
}

/**
 * Checks whether a value is a page component.
 * @param value The value to check.
 */
export function isComponent(value: any): value is Component {
  return value instanceof ComponentImpl;
}
