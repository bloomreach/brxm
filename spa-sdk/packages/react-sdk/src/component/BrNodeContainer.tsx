/*
 * Copyright 2019 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import {
  Container,
  TYPE_CONTAINER_INLINE,
  TYPE_CONTAINER_NO_MARKUP,
  TYPE_CONTAINER_ORDERED_LIST,
  TYPE_CONTAINER_UNORDERED_LIST,
} from '@bloomreach/spa-sdk';
import { BrNodeComponent } from './BrNodeComponent';
import { BrContainerBox, BrContainerInline, BrContainerNoMarkup, BrContainerOrderedList, BrContainerUnorderedList } from '../cms';

export class BrNodeContainer extends BrNodeComponent<Container> {
  protected getMapping() {
    return this.props.component.getType();
  }

  protected fallback() {
    switch (this.props.component.getType()) {
      case TYPE_CONTAINER_INLINE: return <BrContainerInline {...this.props} />;
      case TYPE_CONTAINER_NO_MARKUP: return <BrContainerNoMarkup {...this.props} />;
      case TYPE_CONTAINER_ORDERED_LIST: return <BrContainerOrderedList {...this.props} />;
      case TYPE_CONTAINER_UNORDERED_LIST: return <BrContainerUnorderedList {...this.props} />;
      default: return <BrContainerBox {...this.props} />;
    }
  }
}
