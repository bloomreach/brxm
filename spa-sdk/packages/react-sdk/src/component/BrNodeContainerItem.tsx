/*
 * Copyright 2019 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { ContainerItem } from '@bloomreach/spa-sdk';
import { BrContainerItemUndefined } from '../cms';
import { BrNodeComponent } from './BrNodeComponent';
import { BrProps } from './BrProps';

export class BrNodeContainerItem extends BrNodeComponent<ContainerItem> {
  constructor(props: BrProps<ContainerItem>) {
    super(props);

    this.onUpdate = this.onUpdate.bind(this);
  }

  componentDidMount() {
    this.props.component.on('update', this.onUpdate);
  }

  componentDidUpdate(prevProps: BrProps<ContainerItem>) {
    if (this.props.component !== prevProps.component) {
      prevProps.component.off('update', this.onUpdate);
      this.props.component.on('update', this.onUpdate);
    }
  }

  componentWillUnmount() {
    this.props.component.off('update', this.onUpdate);
  }

  protected getMapping() {
    return this.props.component.getType();
  }

  protected fallback() {
    return <BrContainerItemUndefined {...this.props} />;
  }

  protected onUpdate() {
    this.forceUpdate(() => this.props.page.sync());
  }
}
