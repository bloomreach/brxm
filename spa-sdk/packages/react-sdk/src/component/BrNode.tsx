/*
 * Copyright 2019-2020 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { isContainer, isContainerItem, Component } from '@bloomreach/spa-sdk';
import { BrMeta } from '../meta';
import { BrPageContext } from '../page/BrPageContext';
import { BrComponentContext } from './BrComponentContext';
import { BrNodeContainer } from './BrNodeContainer';
import { BrNodeContainerItem } from './BrNodeContainerItem';
import { BrNodeComponent } from './BrNodeComponent';

interface BrNodeProps {
  component: Component;
}

export class BrNode extends React.Component<BrNodeProps> {
  static contextType = BrPageContext;
  context: React.ContextType<typeof BrPageContext>;

  private renderNode() {
    if (React.Children.count(this.props.children)) {
      return this.props.children;
    }

    const children = this.props.component.getChildren()
      .map((child, index) => <BrNode key={index} component={child} />);

    if (isContainer(this.props.component)) {
      return (
        <BrNodeContainer component={this.props.component} page={this.context!}>
          {children}
        </BrNodeContainer>
      );
    }

    if (isContainerItem(this.props.component)) {
      return (
        <BrNodeContainerItem component={this.props.component} page={this.context!}>
          {children}
        </BrNodeContainerItem>
      );
    }

    return (
      <BrNodeComponent component={this.props.component} page={this.context!}>
        {children}
      </BrNodeComponent>
    );
  }

  render() {
    return(
      <BrComponentContext.Provider value={this.props.component}>
        <BrMeta meta={this.props.component.getMeta()}>
          {this.renderNode()}
        </BrMeta>
      </BrComponentContext.Provider>
    );
  }
}
