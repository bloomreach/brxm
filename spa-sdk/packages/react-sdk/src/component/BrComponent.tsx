/*
 * Copyright 2019 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { BrComponentContext } from './BrComponentContext';
import { BrNode } from './BrNode';

interface BrComponentProps {
  /**
   * The path to a component.
   * The path is defined as a slash-separated components name chain
   * relative to the current component (e.g. `main/container`).
   * If it is omitted, all the children will be rendered.
   */
  path?: string;
}

/**
 * The brXM component.
 */
export class BrComponent extends React.Component<BrComponentProps> {
  static contextType = BrComponentContext;
  context: React.ContextType<typeof BrComponentContext>;

  private getComponents() {
    if (!this.context) {
      return [];
    }
    if (!this.props.path) {
      return this.context.getChildren();
    }

    const component = this.context.getComponent(...this.props.path.split('/'));

    return component ? [component] : [];
  }

  private renderComponents() {
    return this.getComponents().map((component, index) => (
      <BrNode key={index} component={component}>{this.props.children}</BrNode>
    ));
  }

  render() {
    return <>{this.renderComponents()}</>;
  }
}
