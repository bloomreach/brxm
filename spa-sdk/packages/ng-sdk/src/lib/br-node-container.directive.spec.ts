/*
 * Copyright 2020 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Component, Input, NgModule } from '@angular/core';
import { ComponentFixture, TestBed, waitForAsync } from '@angular/core/testing';
import {
  Container,
  TYPE_CONTAINER_BOX,
  TYPE_CONTAINER_INLINE,
  TYPE_CONTAINER_NO_MARKUP,
  TYPE_CONTAINER_ORDERED_LIST,
  TYPE_CONTAINER_UNORDERED_LIST,
} from '@bloomreach/spa-sdk';
import { BrContainerBoxComponent } from './br-container-box/br-container-box.component';
import { BrContainerInlineComponent } from './br-container-inline/br-container-inline.component';
import { BrContainerOrderedListComponent } from './br-container-ordered-list/br-container-ordered-list.component';
import {
  BrContainerUnorderedListComponent,
} from './br-container-unordered-list/br-container-unordered-list.component';
import { BrNodeContainerDirective } from './br-node-container.directive';
import { BrNodeDirective } from './br-node.directive';
import { BrPageComponent } from './br-page/br-page.component';
import { BehaviorSubject } from 'rxjs';

Component({ selector: 'br-container-box', template: '' })(BrContainerBoxComponent);
Component({ selector: 'br-container-inline', template: '' })(BrContainerInlineComponent);
Component({ selector: 'br-container-ordered-list', template: '' })(BrContainerOrderedListComponent);
Component({ selector: 'br-container-unordered-list', template: '' })(BrContainerUnorderedListComponent);

@Component({
  selector: 'br-container-test',
  template: `<a></a>`,
})
class ContainerTestComponent {}

@NgModule({
  declarations: [
    BrContainerBoxComponent,
    BrContainerInlineComponent,
    BrContainerOrderedListComponent,
    BrContainerUnorderedListComponent,
    ContainerTestComponent,
  ],
  entryComponents: [
    BrContainerBoxComponent,
    BrContainerInlineComponent,
    BrContainerOrderedListComponent,
    BrContainerUnorderedListComponent,
    ContainerTestComponent,
  ],
})
class TestModule {}

@Component({ template: '<ng-container [brNodeContainer]="container"></ng-container>' })
class TestComponent {
  @Input() container!: Container;
}

describe('BrNodeContainerDirective', () => {
  let container: jest.Mocked<Container>;
  let node: BrNodeDirective;
  let page: BrPageComponent;
  let component: TestComponent;
  let fixture: ComponentFixture<TestComponent>;

  beforeEach(() => {
    container = {
      getChildren: jest.fn(() => []),
      getType: jest.fn(() => TYPE_CONTAINER_BOX),
      getMeta: jest.fn(() => ({
        clear: jest.fn(),
        render: jest.fn(),
      })),
    } as unknown as typeof container;
    node = {} as typeof node;
    page = {
      mapping: {},
      state: new BehaviorSubject(undefined),
    } as typeof page;
  });

  beforeEach(waitForAsync(() => {
    TestBed.configureTestingModule({
      declarations: [ TestComponent, BrNodeDirective, BrNodeContainerDirective ],
      imports: [ TestModule ],
      providers: [
        { provide: BrNodeDirective, useFactory: () => node },
        { provide: BrPageComponent, useFactory: () => page },
      ],
    })
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(TestComponent);
    component = fixture.componentInstance;
    component.container = container;
  });

  describe('getMapping', () => {
    it('should render a mapped container', () => {
      page.mapping[TYPE_CONTAINER_BOX] = ContainerTestComponent;
      fixture.detectChanges();

      expect(fixture.nativeElement).toMatchSnapshot();
    });

    it('should render box container by default', () => {
      fixture.detectChanges();

      expect(fixture.nativeElement).toMatchSnapshot();
    });

    it('should render an inline container', () => {
      container.getType.mockReturnValue(TYPE_CONTAINER_INLINE);
      fixture.detectChanges();

      expect(fixture.nativeElement).toMatchSnapshot();
    });

    it('should render an ordered list container', () => {
      container.getType.mockReturnValue(TYPE_CONTAINER_ORDERED_LIST);
      fixture.detectChanges();

      expect(fixture.nativeElement).toMatchSnapshot();
    });

    it('should render an unordered list container', () => {
      container.getType.mockReturnValue(TYPE_CONTAINER_UNORDERED_LIST);
      fixture.detectChanges();

      expect(fixture.nativeElement).toMatchSnapshot();
    });

    it('should render a no markup container', () => {
      container.getType.mockReturnValue(TYPE_CONTAINER_NO_MARKUP);
      fixture.detectChanges();

      expect(fixture.nativeElement).toMatchSnapshot();
      expect(container.getChildren).toBeCalled();
    });
  });
});
