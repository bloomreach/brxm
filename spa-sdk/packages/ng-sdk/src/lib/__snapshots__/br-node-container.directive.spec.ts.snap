// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BrNodeContainerDirective getMapping should render a mapped container 1`] = `
<div
  id="root0"
>
  
  <br-container-test>
    <a />
  </br-container-test>
</div>
`;

exports[`BrNodeContainerDirective getMapping should render a no markup container 1`] = `
<div
  id="root5"
>
  
</div>
`;

exports[`BrNodeContainerDirective getMapping should render an inline container 1`] = `
<div
  id="root2"
>
  
  <br-container-inline />
</div>
`;

exports[`BrNodeContainerDirective getMapping should render an ordered list container 1`] = `
<div
  id="root3"
>
  
  <br-container-ordered-list />
</div>
`;

exports[`BrNodeContainerDirective getMapping should render an unordered list container 1`] = `
<div
  id="root4"
>
  
  <br-container-unordered-list />
</div>
`;

exports[`BrNodeContainerDirective getMapping should render box container by default 1`] = `
<div
  id="root1"
>
  
  <br-container-box />
</div>
`;
